### ISSUE TITLE
GetKnownTypes for dynamic known type list

### ISSUE THREAD

**Issue by absinghal** [CoreWCF/CoreWCF#504]
WCF in .Net framework had an overload for ServiceKnownType as follows:

```
[ServiceKnownType("GetKnownTypes", typeof(DataProviderTypes))]
public interface IService
{
}

public class DataProviderTypes
{
    public static IEnumerable<Type> GetKnownTypes(ICustomAttributeProvider provider)
    {
        List<Type> lst = new List<Type>();
        '
        '
        return lst;
    }
}
```

This allowed to dynamically generate list of known types.

Is there any way to achieve this ? If not, when is this planned ? 
Coz this is looking like a showstopper.

Thanks

**Comment by absinghal** [comment]
In the source at [this](https://github.com/CoreWCF/CoreWCF/blob/5408657365fe03de44dc236479bfe0b215871aad/src/CoreWCF.Primitives/src/CoreWCF/Description/TypeLoader.cs#L382) location, following remark is given and relevant code is commented:

"// The named method must take a parameter of ICustomAttributeProvider which isn't available so this can only specify known types by Type"

ICustomAttributeProvider is [available](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.icustomattributeprovider?view=net-6.0) in .Net Standard 2.0, so why cant we just uncomment this code and enable this functionality.

@mconnew will you accept a pull request as per above ?

**Comment by mconnew** [comment]
That comment is a holdover from very early .NET Standard days when that interface wasn't available. It wasn't the reason why that overload wasn't supported though. It was because this code was originally ported for UWP apps running on .NET Native which didn't allow reflection and dynamically calling of methods. This can definitely be re-enable. I'll take a look later to see how much work this will be.

**Comment by mconnew** [comment]
I took a look and here's the steps that need to be done to enable this.
1. Uncomment the disabled properties in ServiceKnownTypeAttribute and fix (just turn into getter only auto properties)
2. Modify CustomAttributeProvider.GetCustomAttributes to also know about ServiceKnownTypeAttribute. This is to provide compat with System.ServiceModel namespaced version of attribute. This will require a new method ConvertFromServiceModelServiceKnownTypeAttribute to be created. It should be easy enough to look at the existing methods to copy the pattern.
3. Uncomment code at top of TypeLoader\<TService\>.GetKnownTypes which handles this case. I believe you can change the second parameter to ICustomAttributeProvider as the abilities of CustomAttributeProvider aren't needed in this method.
4. Add tests to verify all the overloads are working. For this, you can create a base contract interface which won't work without ServiceKnownType added. There should be a negative test validating this is the case. Then for each of the overload variations, create an interface which derives from the base interface and has the ServiceKnownType attribute applied. Tests should also check that System.ServiceModel version of the attribute also works.

I'm going to mark this as help wanted as I'm focusing on getting other features right now.

**Comment by absinghal** [comment]
I can try.

Code changes are [done](https://github.com/absinghal/CoreWCF/commit/e896d2704dfc2f44ea965a5ade3db4ce361a6177) as per above. Only proposed change is that constructor and property setters of ServiceKnownTypeAttribute are internal to enable ConvertFromServiceModelServiceKnownTypeAttribute.

Regarding tests, pl advise whether a new file ServiceKnownTypeTests.cs can be created at [this](https://github.com/absinghal/CoreWCF/tree/main/src/CoreWCF.Primitives/tests) location.

**Comment by absinghal** [comment]
On a related note, during modification of code, it looks like whole of CustomAttributeProvider.cs is a holdover from very early .NET Standard days when ICustomAttributeProvider interface wasn't available. 

I tried to remove it and with some help from referencesource, the solution seems to be compiling fine without any issues.

Should this be committed as well ?

**Comment by absinghal** [comment]
Found some other places where CustomAttributeProvider can be replaced with ICustomAttributeProvider, but looks like CustomAttributeProvider will be anyway required to let both CoreWCF and System.ServiceModel attributes be used, so have left them untouched.

**Comment by mconnew** [comment]
Yes, a new ServiceKnownTypeTests.cs can be created. Feel free to open a draft PR so that you can get some feedback before you are complete if you would like. I'm happy with reducing the usage of CustomAttributeProvider and relegating it to only being used to get custom attributes including the converting System.ServiceModel variations to CoreWCF ones. I think we could get away with turning it into a static class and making the method an extension method on ICustomAttributeProvider. Probably need to rename it otherwise the ICustomAttributeProvider version of the method would get called instead. I'm open to any cleanup you would like to do, or you can make the minimal changes required. It's up to you.
